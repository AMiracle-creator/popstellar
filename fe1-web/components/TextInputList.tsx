import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import RemovableTextInput from 'components/removableTextInput';

/**
 * Component which creates a list of Text input fields which can be deleted individually
 * Returns array with non-empty Values of the text fields
 */

const TextInputList = (props: IPropTypes) => {
  const { onChange } = props;
  const [idCount, setIdCount] = useState(0);
  const [userOptions, setUserOptions] = useState([{ id: -1, value: '' }]);

  const updateParent = (options: { id: number, value: string }[]) => {
    // Gets the distinct options which are not empty ('')
    // Set() keeps the order the same
    const distinctValues = [...new Set(options.map((option) => option.value))].filter((value) => value !== '');
    // Updates the values in the election setup
    onChange(distinctValues);
  };

  const addOption = () => {
    setIdCount(idCount + 1);
    const newOption = { id: idCount, value: '' };
    const newOptions = [...userOptions, newOption];
    setUserOptions(newOptions);
  };

  const updateOption = (id: number, value: string) => {
    const optionIndex = userOptions.findIndex((option) => option.id === id);
    userOptions[optionIndex] = { id: id, value: value };
    setUserOptions(userOptions);
    updateParent(userOptions);
    // If the currently modified textfield is the last in the list
    // then it adds a new text input field
    if (userOptions.filter((option) => option.id > id).length === 0) {
      addOption();
    }
  };

  const removeOption = (id: number) => {
    // This makes sure that when the last textInput is empty, it can't be deleted
    if (userOptions.filter((option) => option.id > id).length !== 0) {
      // This removes the option
      const filteredOptions = userOptions.filter((option) => option.id !== id);
      setUserOptions(filteredOptions);
      updateParent(filteredOptions);
    }
  };

  // Add value to the text so that it doesn't get removed on each re-render
  return (
    <View>
      {userOptions.map((option) => (
        <RemovableTextInput
          onChange={(id: number, text: string) => { updateOption(id, text); }}
          onRemove={removeOption}
          id={option.id}
          value={option.value}
        />
      ))}
    </View>
  );
};

const propTypes = {
  onChange: PropTypes.func.isRequired,
};
TextInputList.propTypes = propTypes;

type IPropTypes = PropTypes.InferProps<typeof propTypes>;

export default TextInputList;
