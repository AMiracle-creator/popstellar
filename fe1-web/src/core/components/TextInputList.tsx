import PropTypes from 'prop-types';
import React, { useMemo } from 'react';
import { View } from 'react-native';

import RemovableTextInput from './RemovableTextInput';

/**
 * Component which creates a list of Text input fields which can be deleted individually
 * Returns array with non-empty Values of the text fields
 * It uses the RemovableTextInput.tsx component to show the individual inputs
 * Input: A function that stores the values that were inputted by the user
 * Output: Array of the unique, nonempty values inputted by the user
 */

const TextInputList = ({ values, onChange, placeholder, testID }: IPropTypes) => {
  const displayedValues = useMemo(() => {
    return [...values, ''];
  }, [values]);

  return (
    <View>
      {displayedValues.map((text, idx) => (
        // FIXME: Do not use index in key
        // eslint-disable-next-line react/no-array-index-key
        <View key={idx.toString()}>
          <RemovableTextInput
            onChange={(newVal) => {
              // remove last element of displayedValues which is always ''
              const newValues = displayedValues.slice(0, -1);
              newValues[idx] = newVal;
              onChange(newValues);
            }}
            onRemove={() => onChange(values.filter((_, id) => idx !== id))}
            value={text || ''}
            isRemovable={idx !== displayedValues.length - 1}
            placeholder={placeholder}
            testID={testID ? `${testID}_option_${idx}` : undefined}
          />
        </View>
      ))}
    </View>
  );
};

const propTypes = {
  values: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  testID: PropTypes.string,
};
TextInputList.propTypes = propTypes;
TextInputList.defaultProps = {
  placeholder: '',
  testID: undefined,
};

type IPropTypes = PropTypes.InferProps<typeof propTypes>;

export default TextInputList;
