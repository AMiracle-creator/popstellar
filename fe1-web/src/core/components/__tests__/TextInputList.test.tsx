import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';

import TextInputList from '../TextInputList';

describe('TextInputList', () => {
  it('renders correctly', () => {
    const { toJSON, getByTestId } = render(
      <TextInputList values={['text  1  ', '   test 2']} onChange={() => {}} testID="x" />,
    );
    const listOption = getByTestId('x_option_0_input');
    fireEvent.changeText(listOption, '    trimmed list   option   ');
    expect(toJSON()).toMatchSnapshot();
  });
});
