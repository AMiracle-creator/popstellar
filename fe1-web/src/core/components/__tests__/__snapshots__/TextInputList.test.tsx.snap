// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInputList renders correctly 1`] = `
<View>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
      testID="x_option_0"
    >
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          editable={true}
          onChangeText={[Function]}
          placeholder=""
          placeholderTextColor="#8E8E8E"
          style={
            [
              {
                "color": "#000",
                "fontSize": 20,
                "lineHeight": 26,
                "marginBottom": 16,
                "textAlign": "left",
              },
              {
                "backgroundColor": "#fff",
                "borderColor": "#fff",
                "borderRadius": 8,
                "borderWidth": 1,
                "flex": 1,
                "padding": 8,
                "width": 50,
              },
            ]
          }
          testID="x_option_0_input"
          value="text  1  "
        />
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "marginBottom": 16,
            "marginLeft": 16,
          }
        }
        testID="x_option_0_remove"
      >
        <RNGestureHandlerButton
          collapsable={false}
          delayLongPress={600}
          enabled={true}
          exclusive={true}
          handlerTag={1}
          handlerType="NativeViewGestureHandler"
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          rippleColor={0}
          touchSoundDisabled={false}
        >
          <View
            accessible={true}
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View>
              {"name":"ios-trash","size":25,"color":"#000"}
            </View>
          </View>
        </RNGestureHandlerButton>
      </View>
    </View>
  </View>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
      testID="x_option_1"
    >
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          editable={true}
          onChangeText={[Function]}
          placeholder=""
          placeholderTextColor="#8E8E8E"
          style={
            [
              {
                "color": "#000",
                "fontSize": 20,
                "lineHeight": 26,
                "marginBottom": 16,
                "textAlign": "left",
              },
              {
                "backgroundColor": "#fff",
                "borderColor": "#fff",
                "borderRadius": 8,
                "borderWidth": 1,
                "flex": 1,
                "padding": 8,
                "width": 50,
              },
            ]
          }
          testID="x_option_1_input"
          value="   test 2"
        />
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "marginBottom": 16,
            "marginLeft": 16,
          }
        }
        testID="x_option_1_remove"
      >
        <RNGestureHandlerButton
          collapsable={false}
          delayLongPress={600}
          enabled={true}
          exclusive={true}
          handlerTag={2}
          handlerType="NativeViewGestureHandler"
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          rippleColor={0}
          touchSoundDisabled={false}
        >
          <View
            accessible={true}
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View>
              {"name":"ios-trash","size":25,"color":"#000"}
            </View>
          </View>
        </RNGestureHandlerButton>
      </View>
    </View>
  </View>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
      testID="x_option_2"
    >
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "row",
          }
        }
      >
        <TextInput
          editable={true}
          onChangeText={[Function]}
          placeholder=""
          placeholderTextColor="#8E8E8E"
          style={
            [
              {
                "color": "#000",
                "fontSize": 20,
                "lineHeight": 26,
                "marginBottom": 16,
                "textAlign": "left",
              },
              {
                "backgroundColor": "#fff",
                "borderColor": "#fff",
                "borderRadius": 8,
                "borderWidth": 1,
                "flex": 1,
                "padding": 8,
                "width": 50,
              },
            ]
          }
          testID="x_option_2_input"
          value=""
        />
      </View>
      <View
        style={
          {
            "marginBottom": 16,
            "marginLeft": 16,
          }
        }
      >
        <View
          style={
            {
              "width": 25,
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;
